### ISSUE TITLE
[SR-14647] `swift build` incorrectly downgrading dependencies in Package.resolved 

### ISSUE THREAD

**Issue by CraigSiemens** [apple/swift-package-manager#4418]
|                  |                 |
   |------------------|-----------------|
   |Previous ID       | SR-14647      |
   |Radar             | rdar://problem/78276237         |
   |Original Reporter | @CraigSiemens      |
   |Type              | Bug    |

Attachment: [Download](https://user-images.githubusercontent.com/2727770/164964189-a35a2753-3809-4725-b565-b377c73b5221.gz)

      
<details>
  <summary>Environment</summary>

`Apple Swift version 5.4 (swiftlang-1205.0.26.9 clang-1205.0.19.55)`  
`Target: x86_64-apple-darwin20.4.0`

</details>

   <details>
  <summary>Additional Detail from JIRA</summary>

   |                  |                 |
   |------------------|-----------------|
   |Votes             | 6         |
   |Component/s       | Package Manager    |
   |Labels            | Bug        |
   |Assignee          | None      |
   |Priority          | Medium      |

   

   md5: 3df0eb42ba22271c1e8a7b759fa20f63

  </details>

**is duplicated by**:  
* [SR-15359](https://bugs.swift.org/browse/SR-15359) `swift package resolve` overwrites `Package.resolved`  

**Issue Description:**

## Introduction

`swift build` will not respect the versions in `Package.swift` and instead change `Package.resolved` to match the dependency versions currently checked out in the `.build` folder.

Seems to be related to SR-10718

## Example project

`SomePackage` is a Swift Package with a git repo, the commits are the following.

`* b6ee888 (update-dependencies) Updated dependencies`  
`* ee23dfe (HEAD -> main) Added dependencies`  
`* 6715449 Initial commit`

`main` is the commit where some dependencies were added

`update-dependencies` is a branch that has updated the dependencies.

## Steps to reproduce

1.  \`cd SomePackage\`

2.  \`git checkout main\`

3.  \`swift build\`

4.  \`git checkout update-dependencies\`

5.  \`swift build\`

## Expected behaviour

It should build the updated dependencies based on what's in `Package.resolved`

## Actual behaviour

The `Package.resolved` is updated to match the version of the dependencies that is currently checked out.

Running `swift package reset` before `swift build` uses the dependency versions from `Package.resolved`

## Additional notes

-   There is a warning cause Apollo isn't added to a target, this is just to speed up building. Adding the dependency as a dependency of a target has no affect.

-   There is a warning about `swift-nio-zlib-support` but I was unable to reproduce the issue with just `swift-nio-zlib-support` or `Starscream` (the subdependecy that pulls in `swift-nio-zlib-support`)

**Comment by leleoveiga** [comment]
same

**Comment by Kondamon** [comment]
This behaviour is horrible when you try to use `git bisect` and have a project with multiple framework dependencies. You are waiting all the time and resetting package caches to get the correctly versions checked out. Is there any workaround for this, or how do you use `git bisect` respecting the `Package.resolved`?

**Comment by neonichu** [comment]
Untested, but `--only-use-versions-from-resolved-file` might help.

**Comment by danyf90** [comment]
> Untested, but --only-use-versions-from-resolved-file might help.

No, it doesn't. The resolve should honor the `Package.resolved` if possible, or use what in `Package.swift` otherwise.
For example, if `Package.resolved` defines version 2.x for a dependency and it is changed in `Package.swift` to 3.x, the Package.resovled should update that dependency, while keeping the other the same, if possible

**Comment by tomassliz** [comment]
I think that #6578 fixes this issue.
